<h2 style="color: darkblue">Borrower Reliability Research<a class="tocSkip"></a></h2>

Our **Client** is the bank's credit department. 

The input data is provided in form of bank statistics about the customers' repayment capacity.

The results of this study will be taken into account when building a **credit scoring model** - a special system that assesses the potential borrower's ability to repay the loan to the bank.

**The aim of the project** is to understand whether a client's marital status and number of children affect their ability to repay the loan on time.

<img src="https://img.freepik.com/free-vector/paying-ability-creditor-and-borrower-shaking-hands-deal-gesture-trust-level-bank-operation-financial-transaction-successful-arrangement_335657-2316.jpg" width="300" height="300" alt="illustration" />

# Data description

* **children** - number of children in the family
* **days_employed** - total employment history in days
* **dob_years** - client's age in years
* **education** - client's education level
* **education_id** - education level identifier
* **family_status** - marital status
* **family_status_id** - marital status identifier
* **gender** - client's gender
* **income_type** - employment type
* **debt** - whether the client has any outstanding loan repayments
* **total_income** - monthly income
* **purpose** - loan purpose

## Data preparation

Read the data from the CSV file into the dataframe and store it in the `data` variable.

In [1]:
# load Pandas library

import pandas as pd

In [2]:
# read file with input data

data = pd.read_csv('./bank_stats.csv')

In [3]:
# display the first 20 lines of the data frame

data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

A preliminary analysis of the data shows that there are anomalous values in the **days_employed** column (negative values) as well as duplicates in the **education** column (same repeating values, but written both using upper and lower case letters).

In [4]:
# output the basic data frame information

data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [5]:
# output the numerical characteristics of the columns

data.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

There are anomalies in our data that do not reflect reality and are due to some error. These anomalies include negative number of days of employment in the `days_employed` column and negative number of children in the `children` column. We will get rid of the anomalous values later on in the data preprocessing phase.

## Data preprocessing

### Identifying and handling the missing values

In [6]:
# output the number of missing values for each column

data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [7]:
# display data frame with empty values in the columns

data[data['days_employed'].isna()]

children  days_employed  dob_years education  education_id  \
12            0            NaN         65   среднее             1   
26            0            NaN         41   среднее             1   
29            0            NaN         63   среднее             1   
41            0            NaN         50   среднее             1   
55            0            NaN         54   среднее             1   
...         ...            ...        ...       ...           ...   
21489         2            NaN         47   Среднее             1   
21495         1            NaN         50   среднее             1   
21497         0            NaN         48    ВЫСШЕЕ             0   
21502         1            NaN         42   среднее             1   
21510         2            NaN         28   среднее             1   

               family_status  family_status_id gender  income_type  debt  \
12          гражданский брак                 1      M    пенсионер     0   
26           женат / замужем                 0      M  госслужащий     0   
29     Не женат / не замужем                 4      F    пенсионер     0   
41           женат / замужем                 0      F  госслужащий     0   
55          гражданский брак                 1      F    пенсионер     1   
...                      ...               ...    ...          ...   ...   
21489        женат / замужем                 0      M    компаньон     0   
21495       гражданский брак                 1      F    сотрудник     0   
21497        женат / замужем                 0      F    компаньон     0   
21502        женат / замужем                 0      F    сотрудник     0   
21510        женат / замужем                 0      F    сотрудник     0   

       total_income                           purpose  
12              NaN                   сыграть свадьбу  
26              NaN                       образование  
29              NaN  строительство жилой недвижимости  
41              NaN  сделка с подержанным автомобилем  
55              NaN                   сыграть свадьбу  
...             ...                               ...  
21489           NaN              сделка с автомобилем  
21495           NaN                           свадьба  
21497           NaN        строительство недвижимости  
21502           NaN  строительство жилой недвижимости  
21510           NaN           приобретение автомобиля  

[2174 rows x 12 columns]

Two columns have missing values. One of them is `days_employed`. We will handle the omissions in this column in the next step. The other column with missing values is `total_income`, which stores income data. 

* **The reason for the omissions** in the original data on income and total employment history of clients could be the borrowers' failure to provide appropriate supporting documents. In this case, it may be appropriate for the bank's internal control department to conduct a sample check of the borrowers' credit files for all relevant documents (their absence has an impact on the credit risk assessment).

Income is most strongly influenced by employment type, so we **fill in the blanks with the median value** for each type in the `income_type` column. The choice to fill the gaps with the median value is due to the fact that this method is not sensitive to outliers and avoids distortions in the distribution of values (keeping the mean and standard deviation in the data).

In [8]:
data['total_income'] = data['total_income'].fillna(data.groupby(['income_type'])['total_income'].transform('median'))

### Processing abnormal values

In [9]:
# replace all negative values with positive values

data['days_employed'] = data['days_employed'].abs()

In [10]:
# for each type of employment, derive the median value of days_employed in days

data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

Two types (безработные as unemployed and пенсионеры as retired) have abnormally high values. It is difficult to correct such values, so we will leave them intact. Moreover, we do not need this column for further study.

In [11]:
# output a list of unique values for the children column

data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

There are two anomalous values in the `children` column. We delete the rows that contain these anomalous values from the `data` dataframe.

In [12]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

In [13]:
# let's check the column again

data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Handling the missing values (continued)

We fill in the gaps in the `days_employed` column with the median values for each employment type `income_type`.

In [14]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

In [15]:
# let's check the column again

data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Changing data types

Here we replace a float data type in the `total_income` column with an integer data type.

In [16]:
data['total_income'] = data['total_income'].astype(int)

### Duplicate handling

In this step we process the duplicates in the **education** column and convert the remaining values to lower case.

In [17]:
data['education'] = data['education'].str.lower()

In [18]:
# display the number of duplicate rows in the data frame

data.duplicated().sum()

71

In [19]:
# remove duplicates

data = data.drop_duplicates()

###  Data categorisation

We create a column `total_income_category` in the data frame with following categories:

- 0-30000 - `'E'`;
- 30001-50000 - `'D'`;
- 50001-200000 - `'C'`;
- 200001-1000000 - `'B'`;
- 1000001 and above - `'A'`.

This will help to better organise the data and perform an in-depth analysis by income category. Categorisation by income could also help with credit and financial planning decisions, as clients at different income levels may have different needs and repayment capacity.

We use our own function called `categorize_income()` and the method `apply()`.

In [20]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [21]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

Now we print out a list of unique borrowing targets from the **purpose** column.

In [22]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

Now we define a function that uses the data in the `purpose` column to form a new column `purpose_category` that includes the following categories:

- `'операции с автомобилем'` as `'motor vehicle operations'`,
- `'операции с недвижимостью'` as `'real estate transactions'`,
- `'проведение свадьбы'` as `'wedding venue'`,
- `'получение образования'` as `'getting an education'`.

Let's use our own function named `categorize_purpose()` and method `apply()`. Next, we will examine the data in the `purpose` column and determine which substrings will help us correctly identify the category.

In [23]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'motor vehicle operations'
        elif 'жил' in row or 'недвиж' in row:
            return 'real estate transactions'
        elif 'свад' in row:
            return 'wedding venue'
        elif 'образов' in row:
            return 'getting an education'
    except:
        return 'no category'

In [24]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Data analysis

#### Analysis whether there is a correlation between the number of children and repayment on time.

In [25]:
# use a function to format the non-return rate values as a percentage with two decimal places

def percent(x):
    return "{0:.2%}".format(x)

# use the function to build a summary table to summarize the number of loans, debtors and default rates 
# by each category

def dependence(category):
    category_debt = data.pivot_table(index=category, values='debt', aggfunc=['count', 'sum', 'mean'])
    
    # let's rename the column names for convenience
    
    category_debt.columns = ['Total borrowers', 'Total debtors', 'Share of debtors']
    
    # call the function with the right format
    
    category_debt['Share of debtors'] = category_debt['Share of debtors'].apply(percent)
    
    return category_debt.sort_values(by='Share of debtors', ascending=False)

In [26]:
dependence('children')

Total borrowers  Total debtors Share of debtors
children                                                 
4                      41              4            9.76%
2                    2052            194            9.45%
1                    4808            444            9.23%
3                     330             27            8.18%
0                   14091           1063            7.54%
5                       9              0            0.00%

**Conclusion:** 
according to the data obtained, the available sample is unbalanced - the values of the number of clients broken down by the number of children in the family are not comparable. Let's allocate 3 following groups to correct conclusions in the comparative analysis : "no children", "1 child in the family" and "2 children in the family". 
Based on the data of these three groups, it can be noted that the presence of children in the family increases the probability of overdue debts. The most risky borrowers are clients with 2 children (risk of default - 9.45%), which is explained by the large amount of expenses per family member compared to the rest of the sample. The most "attractive" borrowers are the clients without children (in the absence of additional expenses this category of clients is more likely to fulfill their obligations to the bank). 

Thus, the **correlation between the number of children and loan repayment on time has been established.**

#### Analysis of the relationship between marital status and loan repayment on time.

In [27]:
# create a dictionary to replace the values with English translations

family_status_dict = {
    'Не женат / не замужем': 'Single',
    'в разводе': 'Divorced',
    'вдовец / вдова': 'Widowed',
    'гражданский брак': 'Civil marriage',
    'женат / замужем': 'Married'
}

data['family_status'] = data['family_status'].replace(family_status_dict)

In [28]:
dependence('family_status')

Total borrowers  Total debtors Share of debtors
family_status                                                  
Single                     2796            273            9.76%
Civil marriage             4134            385            9.31%
Married                   12261            927            7.56%
Divorced                   1189             84            7.06%
Widowed                     951             63            6.62%

**Conclusion:** based on the above grouping, it can be noted that **family status also influences loan repayment**. Widowers and divorcees are more disciplined borrowers - we can assume that this group is older relative to the other categories and therefore has fewer unplanned expenses, needs and more experience in rationally distributing their income. Married clients are more diligent about repayment than single clients. This may be due to the fact that spouses may be co-borrowers (jointly liable under the loan agreement).

#### Analysis of the relationship between income level and loan repayment on time.

In [29]:
dependence('total_income_category')
# where A - very high income (over 1 million roubles)
# B - high income (from 200 to 1 million roubles)
# C - upper middle income (from 50 to 200 thousand roubles)
# D - middle income (from 30 to 50 thousand roubles)
# E - low income (up to 30 thousand roubles)

Total borrowers  Total debtors Share of debtors
total_income_category                                                 
E                                   22              2            9.09%
C                                15921           1353            8.50%
A                                   25              2            8.00%
B                                 5014            354            7.06%
D                                  349             21            6.02%

**Conclusion:** 
from the obtained data we can see that the number of clients by category is highly unbalanced. If the benchmarking would be done on the basis of such a sample, the conclusions may not be correct and may not correspond to reality. Using the results for practical purposes may lead to negative consequences - a trustworthy client may be recognised as a 'bad' client and vice versa. Thus, it is useful to compare multiple samples: B "high income" and C "above average income".

The results of this comparison confirmed the obvious fact that clients with "high" income make fewer late payments than people with less income who fall into the "upper-middle income" category.
With this in mind, we confirm the **dependence of income on delinquency risk**.

#### Analysis of the impact of the purpose of the loan on repayment on time

In [30]:
dependence('purpose_category')

Total borrowers  Total debtors Share of debtors
purpose_category                                                         
motor vehicle operations             4279            400            9.35%
getting an education                 3988            369            9.25%
wedding venue                        2313            183            7.91%
real estate transactions            10751            780            7.26%

**Conclusion:** the risk of default is higher for borrowers on car loans and education loans. Borrowers with loans for real estate transactions are the most disciplined - it is likely that clients are more responsible about their own homes. The indicator may be influenced by borrowers who have taken out a property loan and are repaying the debt from the proceeds of renting it out. Given the above, we can conclude **that there is an impact of the purpose of the loan on repayment.**

### General conclusion

Based on the conducted analysis, it can be concluded that **all factors considered have an impact on loan repayment on time**:
 1. There is a correlation between the number of children and loan repayment on time. Given the imbalance in the sample, 3 groups were considered: "no children", "1 child in the family", "2 children in the family". The results of the analysis showed that the higher the number of children, the higher the risk of default: **the riskiest** borrowers are the clients **with 2 children** (default risk of 9.4%), **in second place are families with 1 child** (9.2%) and **the most "attractive"** borrowers are clients **without children** (7.5%).
 2. There is a correlation between marital status and loan repayment on time. The survey results show that **family clients and those who have been married at least once are less likely to default than single**: the average debtor rate in the "bachelor" group is 9.8%, common-law marriage is 9.3%, registered marriage is 7.6%, divorced - 7%, widowed - 6.6%.
 3. There is a correlation between income level and loan repayment on time. The results of the analysis confirmed that **borrowers with high income are more disciplined** in repayment than **borrowers with above average income** (the share of debtors is 7% and 8.5% respectively). Because of the unbalanced performance of the other groups, only multiple samples B 'high income' and C 'upper-middle income' were compared to reveal the relationship.
 4. Different purposes of the loan have their own impact on its repayment on time - **the most risky loans** are **autocredit** and **loans for education** (share of debtors - 9.3% and 9.2%, respectively). Clients who has taken advantage of **loans for real estate** allow **fewer overdue payments** (default risk - 7.5%).

At the same time, to improve scoring, we recommend increasing the set of factors for risk assessment. For example, include such factors as client's age, credit history (including overdue debts in other banks), availability of accounts with the bank. Also it is necessary to pay attention to availability of all necessary supporting documents (documents from workplace, confirming the client's solvency) when considering the client's loan application. The bank's credit department is advised to check the completeness and relevance of the documents in the borrowers' credit files on an ongoing basis.